include/pybind11/detail/internals.h

```diff
@@ -510,11 +510,19 @@ struct local_internals {
 };
 
 /// Works like `get_internals`, but for things which are locally registered.
+/// There is one variable storing the local internals for each module, but every embedded module
```

Comment:
I'm sorry if it does not sound clear, English is not my native language. What I meant is that since the static local variable of `get_local_internals` does not get dynamically linked when the extension module is loaded by the interpreter, you end up with one definition of the locals for the interpreter, and one definition for each non-embedded modules.

Reply:
I think in that case my suggested wording is clearer. But I'm not a native speaker, too!
Let's see what @Skylion007 thinks.

Reply:
Sorry I'm having serious 2nd thoughts.

E.g. at Google, Python and all extension modules for a given binary (or test) are linked together statically. Each binary or test is hermetic.

Others may use `RTLD_GLOBAL`. Probably rare, but maybe in certain environment that's what's needed?

What does this PR mean in those situations? I'm not sure. **Does someone else have a clear understanding of how the `m_clear` functionality plays out across all environments?**

Guessing a lot: I think to be completely safe, we'd need some kind of lookup table, to keep track of which `module_def` pointers are associated with a given `local_internals` pointer, and if more than one, we cannot safely `clear_local_internals` until the last one disappears.

Or we could have `static std::unordered_map<module_def *, local_internals>` with all the required logic to register & clear types given a `module_def` pointer, but search for types over all keys of the map.

It seems pretty tricky to get the `m_clear` functionality correct in all environments.

Given those uncertainties, and that pybind11 existed to this point without using `m_clear`, maybe this should be an opt-in, e.g. via some `#define`, or runtime option? That would be a safer and less sudden way to find out what's possible and what not.